### ISSUE TITLE
Use WebViewCompat.addWebMessageListener instead of @JavascriptInterface to prevent origin spoofing

### ISSUE THREAD

**Issue by tian000** [react-native-webview/react-native-webview#3439]
Currently if the web app redirects quickly after doing `window.ReactNativeWebview.postMessage` you could receive a spoofed origin on the react native side.

We should use WebViewCompat.addWebMessageListener to prevent origin spoofing.

This is critical for security sensitive apps like crypto wallets. 

I can put up a PR for this if maintainers a willing to accept a fix for this issue.

**Comment by otech47** [comment]
could you elaborate further how an attacker would exploit this? would an attacker aiming to spoof a domain need control over the web app that is redirecting quickly to then trick a user into thinking they are on google.com when really they are using maliciouswebsite.com?

**Comment by tian000** [comment]
Exactly the malicious attacker controls the malicious site (usually a phishing site) and tricks the user into thinking it is a legitimate site via fast redirects.

It works something like this:
1. User lands on malicioussite.com
2. malicious site redirects to `google.com`, and sends a request to the mobile app with a delay e.g. 
    ```
    setTimeout(() => window.ReactNativeWebview.postMessage("I am google"), 5)
    ```
3. The mobile app thinks that the request is coming from google.com even though it is coming from malicioussite.com

**Comment by tian000** [comment]
@otech47 what do you think?

**Comment by otech47** [comment]
> @otech47 what do you think?

I didn't attempt to reproduce. I just wanted to confirm the low risk to a use case in which the user specifies which URL they access so the security impact for that is minimal imo

But I'm sure it is more dangerous in plenty of other use cases and should be fixed

**Comment by rafinskipg** [comment]
Yeah, we see it being reported several times as a very dangerous origin spoofing issue. Since it can fake the website that is emitting the messages and make the users believe that a legitimate site is requesting them to sign some transaction etc

**Comment by tian000** [comment]
@otech47 i've put up a PR for the fix here: https://github.com/react-native-webview/react-native-webview/pull/3609

**Comment by tian000** [comment]
Fixed in https://github.com/react-native-webview/react-native-webview/pull/3609
